server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java

```diff
@@ -0,0 +1,660 @@
+                    // In a new currentTerm, we cannot compare the persisted metadata's lastAcceptedVersion to those in the new state, so
+                    // it's simplest to write everything again.
+                    overwriteMetaData(clusterState);
+                } else {
+                    // Within the same currentTerm, we _can_ use metadata versions to skip unnecessary writing.
+                    updateMetaData(clusterState);
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+
+            commit(currentTerm, clusterState.version());
+            lastAcceptedState = clusterState;
+        }
+
+        /**
+         * Update the persisted metadata to match the given cluster state by removing any stale or unnecessary documents and adding any
+         * updated documents.
+         */
+        private void updateMetaData(ClusterState clusterState) throws IOException {
+            assert lastAcceptedState.term() == clusterState.term();
+            logger.trace("currentTerm [{}] matches previous currentTerm, writing changes only", clusterState.term());
+
+            try (ReleasableDocument globalMetaDataDocument = makeGlobalMetaDataDocument(clusterState)) {
+                for (MetaDataIndexWriter metaDataIndexWriter : metaDataIndexWriters) {
+                    metaDataIndexWriter.deleteGlobalMetaData();
+                    metaDataIndexWriter.addGlobalMetaData(globalMetaDataDocument.getDocument());
+                }
+            }
+
+            final Map<String, Long> indexMetaDataVersionByUUID = new HashMap<>(lastAcceptedState.metaData().indices().size());
+            for (ObjectCursor<IndexMetaData> cursor : lastAcceptedState.metaData().indices().values()) {
+                final IndexMetaData indexMetaData = cursor.value;
+                final Long previousValue = indexMetaDataVersionByUUID.putIfAbsent(indexMetaData.getIndexUUID(), indexMetaData.getVersion());
+                assert previousValue == null : indexMetaData.getIndexUUID() + " already mapped to " + previousValue;
+            }
+
+            for (ObjectCursor<IndexMetaData> cursor : clusterState.metaData().indices().values()) {
+                final IndexMetaData indexMetaData = cursor.value;
+                final Long previousVersion = indexMetaDataVersionByUUID.get(indexMetaData.getIndexUUID());
+                if (previousVersion == null || indexMetaData.getVersion() != previousVersion) {
+                    if (previousVersion != null) {
+                        logger.trace("overwriting metadata for [{}], changing lastAcceptedVersion from [{}] to [{}]",
+                            indexMetaData.getIndex(), previousVersion, indexMetaData.getVersion());
+                        for (MetaDataIndexWriter metaDataIndexWriter : metaDataIndexWriters) {
+                            metaDataIndexWriter.deleteIndexMetaData(indexMetaData.getIndexUUID());
```

Comment:
should we just use `updateDocument` here in all cases instead of optimizing this with `addDocument`? I think it might not be worth optimizing this and will make the code simpler.

Reply:
Sure, see 7fa29f96ad9c2d92cfff9f38a53026f0af5917d5.